### DISCUSSION TITLE
Object files are not link

### DISCUSSION THREAD

**Discussion by BeardPower** [unikraft/unikraft#269] [Q&A]
Hi,

I have this Makefile.uk, but the object file is not linked to the application. It is picked up (make print-obj), but the linker does not link it (I checked with make V=1). Did I miss something?

$(eval $(call addlib,apphelloworld))

APPHELLOWORLD_SRCS-y += $(APPHELLOWORLD_BASE)/main.c
APPHELLOWORLD_OBJS-y += $(APPHELLOWORLD_BASE)/hw.o

**Comment by razvand** [comment]
Hi, @BeardPower.

I tested this[1] on my end and it works.

Do the following to build and test it:
* build `test.o` by running the `build-test` script
* run `make menuconfig` and simply save the configuration (it will load the `Config.uk` file)
* run `make`
* run the `linuxu` build with `./build/app-empty_linuxu-x86_64`
* run the `kvm` build with `./qemu-guest -k build/app-empty_kvm-x86_64`

See if you can adapt that for your use case and figure out why yours doesn't work.

Or you could provide a repository link so we can check it out.

[1] https://github.com/cs-pub-ro/app-empty (branch `test-obj`).

**Reply by BeardPower** [reply]
Hi @razvand,

Thank you for your fast response! Sorry, I mixed some things. Using an object file works fine, but not using an archive:
APPEMPTY_OBJS-y += $(APPEMPTY_BASE)/mylib.a

You can find my repository with the test here: https://github.com/BeardPower/app-empty/tree/test-obj

**Reply by Rochet2** [reply]
I believe I run into same issue.
Linking a main.cpp that contains a main function and main.o that contains main function fails with an error as expected.
But linking main.a that contains the main.o does not generate an error, which means that the archive did not get linked.

**Reply by Rochet2** [reply]
Created an issue here https://github.com/unikraft/unikraft/issues/350

**Comment by Rochet2** [comment]
It seems that the following works fine for me in my test, so it likely works for you as well:
```
APPEMPTY_ALIBS-y += $(APPEMPTY_BASE)/mylib.a
```

The ALIBS variable seems to be meant for exactly this purpose, but is not documented as far as I see or the existing documentation has a mistake.
Remember to build the project from scratch (`kraft clean && kraft build`) so that old remains of previous builds do not affect the fixed build setup.

See more info about this in https://github.com/unikraft/unikraft/issues/350

**Comment by razvand** [comment]
Everyone, since we are already (viciously) using [the Unikraft Discord server](https://bit.ly/UnikraftDiscord) for discussions, meetings, brainstorming and jokes (mostly bad, we do what we can), we will close the Discussions tab. We will do this on Friday, March 31, 2023.

Please move any discussions or questions on [the Unikraft Discord server](https://bit.ly/UnikraftDiscord). See you all there! :handshake:
